### ISSUE TITLE
中间层Loss，会去更新后面网络的参数吗

### ISSUE THREAD

**Issue by DvHuang** [airaria/TextBrewer#86]
假如蒸馏一个4层的学生模型，以第二层为学习目标，那么该Loss，产生的梯度，会去更新第三层第四层的参数吗？
看代码好像直接是total_loss.backward()，这样是不是用中间层的loss去更新后后面层的参数

**Comment by airaria** [comment]
中间层的loss只会影响它下面的层的参数。
以第二层为学习目标，该loss产生的梯度只会影响embedding和1、2层
在2021年11月22日 ***@***.***> 写道：

假如蒸馏一个4层的学生模型，以第二层为学习目标，那么该Loss，产生的梯度，会去更新第三层第四层的参数吗？
看代码好像直接是total_loss.backward()，这样是不是用中间层的loss去更新后后面层的参数

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by DvHuang** [comment]
感谢解答，您说的这个是说pytorch会自动区分去更新哪些层的参数吗，代码里total_loss 是一个数值吧，执行total_loss.backward()的时候，就没办法区分这里面的loss来自哪里了吧

**Comment by airaria** [comment]
PyTorch会根据得到loss的计算图去更新涉及到的参数。
total_loss是一个标量，PyTorch背后的机制是通过记录total_loss的计算流程可以区分total_loss中的每一项是来自哪，但对于用户没必要区分。
在2021年11月22日 ***@***.***> 写道：

感谢解答，您说的这个是说pytorch会自动区分去更新哪些层的参数吗，代码里total_loss 是一个数值吧，执行total_loss.backward()的时候，就没办法区分这里面的loss来自哪里了吧

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by DvHuang** [comment]
您好，刚才把参数输出了一下，可能pytorch并没有区分total_loss 中的来源
具体做法：
在random token example 这个例子中，将源代码中的学习目标只保留如下Loss：
        {'layer_T': 3, 'layer_S': 1, 'feature': 'hidden', 'loss': 'hidden_mse', 'weight': 1},

然后在optimizer.step() 前后分别输出第三层（序号为2）key 矩阵，得到结果如下：
before optimizer params:bert.encoder.layer.2.attention.self.key.weight:
tensor([[-0.0286,  0.0178, -0.0051,  ..., -0.0168, -0.0497,  0.0153],
        [-0.0141,  0.0190, -0.0041,  ..., -0.0192,  0.0163,  0.0300],
        [ 0.0352,  0.0099,  0.0212,  ...,  0.0391, -0.0078,  0.0012],
        ...,
        [ 0.0285, -0.0063,  0.0167,  ..., **-0.0014**,  0.0413,  0.0225],
        [ 0.0044, -0.0038, -0.0184,  ...,  0.0024, -0.0033,  0.0158],
        [-0.0131,  0.0142, -0.0007,  ..., -0.0068,  0.0039,  0.0142]],
       requires_grad=True)
after  optimizer params:bert.encoder.layer.2.attention.self.key.weight:
tensor([[-0.0286,  0.0178, -0.0051,  ..., -0.0168, -0.0497,  0.0153],
        [-0.0141,  0.0190, -0.0041,  ..., -0.0192,  0.0163,  0.0300],
        [ 0.0352,  0.0099,  0.0212,  ...,  0.0391, -0.0078,  0.0012],
        ...,
        [ 0.0285, -0.0063,  0.0167,  ..., **-0.0015**,  0.0413,  0.0225],
        [ 0.0044, -0.0038, -0.0184,  ...,  0.0024, -0.0033,  0.0158],
        [-0.0131,  0.0142, -0.0007,  ..., -0.0068,  0.0039,  0.0142]],

可以看到两个矩阵，加粗位置的参数发生了变化
